### DISCUSSION TITLE
Memory usage

### DISCUSSION THREAD

**Discussion by wwatu1337** [inventree/InvenTree#3505] [Q&A] [answered]
I'm interested in what is your memory usage, with supervisor and default config I see about 450mb memory usage without any traffic, seems a bit high.

**Comment by miggland** [comment]
I'm on docker, total of four containers (excluding Traefik proxy), using about 750 MB of memory.

**Comment by SchrodingersGat** [comment]
@wwatu1337 are you running the debug server or gunicorn?

**Reply by wwatu1337** [reply]
Gunicorn

**Comment by matmair** [comment]
@wwatu1337 that is somewhat normal for Django apps. In the normal setup it is running 1 webserver, 1 app-server, 4 worker + you propably have a database on there too. The good news is that it scales pretty well on kubernetes if you adjust a few things.

**Reply by wwatu1337** [reply]
I adjusted gunicorn workers to just 1, this made the server process take around 130mb of ram, with 3 workers around 300mb, so this is expected behaviour, with the server stopped so only the cluster process was running, it takes a whopping 800-900mb ram, this still seems a bit high for me, is there any way to limit this, I can see that I have about 8 qcluster processes taking around 100mb each, so 800mb in total
![image](https://user-images.githubusercontent.com/58936586/196742492-ece52498-c79e-4e8c-8b6c-290b4cd0b162.png)
![image](https://user-images.githubusercontent.com/58936586/196744004-cc9d850b-955f-4620-a489-96978e43ef67.png)

**Reply by matmair** [reply]
@wwatu1337 having only 1 gunicorn worker will significantly influence the user experience as that means only one request accessing the database is possible at a given time. We have views that run up to 9 webrequest simultaneously so these views will have to waterfall.

8 cluster processes is a lot and the ram seems also high. How are you deploying and have you configured workers in any way?

**Reply by wwatu1337** [reply]
This was more for a debugging purposes to see where the memory usage was coming with. The problem seems to be with 8 cluster processes.

I have bare bones supervisor production setup with the default supervisor conf.

`supervisorctl status` returns
![image](https://user-images.githubusercontent.com/58936586/196813882-645f3241-42b8-43ef-978d-e9a36a73c5e5.png)

`pstree -gp` output
![image](https://user-images.githubusercontent.com/58936586/196813828-0a0c39c5-f464-4553-b612-9c0976f30819.png)

**Reply by matmair** [reply]
@wwatu1337 you can tune the number of worker processes in the supervisor setup or with the settings - that might help reduce the footprint.

**Reply by wwatu1337** [reply]
Default of numprocs is 1, as supervisorctl status returns only 1 cluster process is started, these are child processes spawned by manage.py, or is there any way in configurating those?

I have now further looked into this, I've set up a local docker env, while the server process has similar memory usage the worker process is very different, the memory usage is only around 150-200mb, this seems way more reasonable, it still spawns 8 processes as seen on the image, but the memory usage is very different, so something must definately be off, but out of ideas what that might be.

![image](https://user-images.githubusercontent.com/58936586/196927062-3d529e3f-7490-4b7a-9a02-23941b9d576e.png)
![image](https://user-images.githubusercontent.com/58936586/196927102-21986c6b-51f1-4d87-a83a-0029d12f6505.png)

**Reply by matmair** [reply]
@wwatu1337 that number is way more in line with what I see. You can set the number of workers in the config.yaml.

**Reply by wwatu1337** [reply]
with only 1 worker, the memory usage drops about 200mb, but I would be more interested in finding out what causes the huge memory usage, any suggestions on what to try/look for?

**Reply by matmair** [reply]
@wwatu1337 tbh I find memory profiling with Django somewhat cumbersome and never bothered up to now as it wasn't a real problem. There is an extension out there that can be used with django-debug-toolbar but not sure if that will work with the worker.

You might also be able to hook into the python runtime with New Relic and analyze that way. I have not done that my self, colleagues have told me about the option.

**Comment by matmair** [comment]
@wwatu1337 could you please select an answer or let us know what you want further answered?

**Reply by wwatu1337** [reply, marked as answer]
sorry for late reply, just moved from bare metal to package manager installation, memory usage out of the box is around 2GB, setting gunicorn workers to 1 via
`inventree config:set GUNICORN_CMD_ARGS='--workers=1 --access-logfile -'`
brings the memory down to acceptable 1GB
